asserts/ifacedecls.go

```diff
@@ -220,7 +222,256 @@ func compileAttributeConstraints(constraints interface{}) (*AttributeConstraints
+)
+
+func compilePlugConnectionConstraints(interfaceName string, entry string, constraints interface{}) (*PlugConnectionConstraints, error) {
+	context := fmt.Sprintf("%s in plug rule for interface %q", entry, interfaceName)
+	cMap, shortcut, err := checkMapOrShortcut(context, constraints)
+	if err != nil {
+		return nil, err
+	}
+	if cMap == nil {
+		if shortcut {
+			return &PlugConnectionConstraints{PlugAttributes: AlwaysMatchAttributes, SlotAttributes: AlwaysMatchAttributes}, nil
+		}
+		return &PlugConnectionConstraints{PlugAttributes: NeverMatchAttributes, SlotAttributes: NeverMatchAttributes}, nil
+	}
+	plugConnCstrs := &PlugConnectionConstraints{}
+	defaultUsed := 0
+	for _, field := range plugIDConstraints {
+		// TODO: check these further against regexps?
+		l, err := checkStringListInMap(cMap, field, fmt.Sprintf("%s in %s", field, context))
+		if err != nil {
+			return nil, err
+		}
+		if l == nil {
+			defaultUsed++
+		}
+		plugConnCstrs.setIDConstraints(field, l)
+	}
+	for _, field := range attributeConstraints {
+		cstrs := AlwaysMatchAttributes
+		v := cMap[field]
+		if v != nil {
+			var err error
+			cstrs, err = compileAttributeConstraints(cMap[field])
+			if err != nil {
+				return nil, fmt.Errorf("cannot compile %s in %s: %v", field, context, err)
+			}
+		} else {
+			defaultUsed++
+		}
+		plugConnCstrs.setAttributeConstraints(field, cstrs)
+	}
+	if defaultUsed == len(attributeConstraints)+len(plugIDConstraints) {
+		return nil, fmt.Errorf("%s must specify at least one of %s, %s", context, strings.Join(attributeConstraints, ", "), strings.Join(plugIDConstraints, ", "))
+	}
+	return plugConnCstrs, nil
+}
+
+var (
+	defaultAllowRule = map[string]interface{}{
+		"allow-installation":    "true",
+		"allow-connection":      "true",
+		"allow-auto-connection": "true",
+		"deny-installation":     "false",
+		"deny-connection":       "false",
+		"deny-auto-connection":  "false",
+	}
+
+	denyRule = map[string]interface{}{
+		"allow-installation":    "false",
+		"allow-connection":      "false",
+		"allow-auto-connection": "false",
+		"deny-installation":     "true",
+		"deny-connection":       "true",
+		"deny-auto-connection":  "true",
+	}
+
+	installationSubrules = []string{"allow-installation", "deny-installation"}
+	connectionsSubrules  = []string{"allow-connection", "deny-connection", "allow-auto-connection", "deny-auto-connection"}
+)
+
+func compilePlugRule(interfaceName string, rule interface{}) (*PlugRule, error) {
+	context := fmt.Sprintf("plug rule for interface %q", interfaceName)
+	rMap, shortcut, err := checkMapOrShortcut(context, rule)
+	if err != nil {
+		return nil, err
+	}
+	if rMap == nil {
+		if shortcut {
+			rMap = defaultAllowRule
+		} else {
+			rMap = denyRule
+		}
+	}
+	plugRule := &PlugRule{
+		Interface: interfaceName,
+	}
+	defaultUsed := 0
+	// installation subrules
+	for _, subrule := range installationSubrules {
+		v := rMap[subrule]
+		if v == nil {
+			v = defaultAllowRule[subrule]
```

Comment:
I had some difficulty reading this part of the logic, and I think the problem is the name of this variable. This mixes the idea of "default" and "allow", when some of those rules are themselves named allow, and others are deny. so the "default allow of deny".. (brain explodes)

Perhaps "defaultOutcome" and "invertedOutcome"?